### PR TITLE
Hive Meta Store impersonation access

### PR DESCRIPTION
**Pull request by BlueStalker**
This commits squash the original commits from PR
https://github.com/prestodb/presto/pull/13699
which includes the follow commits:

HMS impersonation access
refactoring to use HMS Authentication Module
add Config for multiple hms instances
Update HMS memory settings
address review comments

```
== RELEASE NOTES ==

General Changes
* N/A

Hive Changes
* add hive metastore impersonation access
* add multiple hive metastore instances support and metrics
* Code refactoring
```
If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by BlueStalker** [comment]
@arhimondr This comes from https://github.com/prestodb/presto/pull/13699 and let's make another round of code reviews. Thanks. Also, FYI, @zhenxiao

**Comment by arhimondr** [comment]
@BlueStalker  Thank you, will start reviewing it soon.

File: pom.xml [hunk 1]
```diff
@@ -1450,6 +1450,12 @@
                 <version>1.3.5-4</version>
             </dependency>
 
+            <dependency>
```
Comment (arhimondr) [inline pom.xml:1453]: We don't use `ApacheCommons` in Presto. Please use Guava instead.

File: presto-docs/src/main/sphinx/connector/hive-security.rst [hunk 2]
```diff
@@ -131,6 +131,8 @@ Property Name                                      Description
                                                    to the Hive metastore service.
 
 ``hive.metastore.client.keytab``                   Hive metastore client keytab location.
+``hive.metastore.impersonation.enabled``           Enable metastore end-user impersonation.
+``hive.metastore.impersonation.user``              Default impersonation user when communicating with Hive Metastore
```
Comment (arhimondr) [inline presto-docs/src/main/sphinx/connector/hive-security.rst:135]: In what cases we fallback to the default user? It feels like if impersonation is enabled we should never contact the metastore with "default".

File: presto-hive-hadoop2/bin/run_hive_tests.sh [hunk 3]
```diff
@@ -7,6 +7,11 @@ set -euo pipefail -x
 cleanup_docker_containers
 start_docker_containers
 
+# restart HMS to pickup memory settings
+exec_in_hadoop_master_container cp /etc/hadoop/conf/hive-env.sh /etc/hive/conf/hive-env.sh
```
Comment (arhimondr) [inline presto-hive-hadoop2/bin/run_hive_tests.sh:11]: These changes should go directly to the docker containers. Here is the repository: https://github.com/prestodb/docker-images

File: presto-hive-hadoop2/conf/files/hive-env.sh [hunk 4]
```diff
@@ -0,0 +1 @@
+export HADOOP_OPTS="$HADOOP_OPTS -Xmx1024m"
```
Comment (arhimondr) [inline presto-hive-hadoop2/conf/files/hive-env.sh:1]: new line

File: presto-docs/src/main/sphinx/connector/hive-security.rst (hunk 2 above)
Comment (BlueStalker) [reply presto-docs/src/main/sphinx/connector/hive-security.rst:135]: MetastoreHiveStatisticsProvider#getPartitionsStatistics, it gets the paritition stats.

**Comment by arhimondr** [comment]
@BlueStalker The global metastore cache can be enabled with https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreClientConfig.java#L35.

Also It doesn't feel right to wrap with `doAs` at the `SemiTransactionalHiveMetastore` level. The `doAs` will only work for ThriftHiveMetastoreClient.java. Wrapping it with `doAs` actually breaks encapsulation. Other metastore implementation may have their own means of providing user. It feels like the information about identity has to be passed all the way down to the `ThriftHiveMetastoreClient`, and then the `ThriftHiveMetastoreClient` can implement authentication by applying `doAs`, as this is very `ThriftHiveMetastoreClient` specific.

**Pull request closed without merging**